[mlir][Ptr] Add the `MemorySpaceAttrInterface` interface and dependencies.

fabianmcg opened the pull request:
This patch introduces the `MemorySpaceAttrInterface` interface. This interface is responsible for handling the semantics of `ptr` operations.

For example, this interface can be used to create read-only memory spaces,  making any other operation other than a load a verification error, see `TestConstMemorySpaceAttr` for a possible implementation of this concept.

This patch also introduces Enum dependencies `AtomicOrdering`, and `AtomicBinOp`, both enumerations are clones of the Enums with the same name in the LLVM Dialect.

Also, see:
- [[RFC] `ptr` dialect & modularizing ptr ops in the LLVM dialect](https://discourse.llvm.org/t/rfc-ptr-dialect-modularizing-ptr-ops-in-the-llvm-dialect/75142) for rationale.
- https://github.com/llvm/llvm-project/pull/73057 for a prototype implementation of the full change.

**Note: Ignore the first commit, that's being reviewed in https://github.com/llvm/llvm-project/pull/86860 .**

fabianmcg committed:
[mlir][Ptr] Add the `MemorySpaceAttrInterface` interface and dependencies.

This patch introduces the `MemorySpaceAttrInterface` interface. This
interface is responsible for handling the semantics of `ptr` operations.
For example, this interface can be used to create read-only memory spaces,
making any other operation other than a load a verification error, see
`TestConstMemorySpaceAttr` for a possible implementation of this concept.

This patch also introduces Enum depedencies `AtomicOrdering`, and `AtomicBinOp`,
both enumerations are clones of the Enums with same name in the LLVM Dialect.

Dinistro commented on mlir/include/mlir/Dialect/Ptr/IR/MemorySpace.h:
```diff
@@ -0,0 +1,161 @@
+class MemorySpace {
+public:
+  MemorySpace() = default;
+  MemorySpace(std::nullptr_t) {}
+  MemorySpace(MemorySpaceAttrInterface memorySpace)
+      : memorySpaceAttr(memorySpace), memorySpace(memorySpace) {}
+  MemorySpace(Attribute memorySpace)
```
Nit: Do we want implicit conversion constructors? Especially from `Attribute`?

fabianmcg replied:
I can make it explicit. I left it implicit for convenience.

Dinistro commented on mlir/include/mlir/Dialect/Ptr/IR/MemorySpace.h:
```diff
@@ -0,0 +1,161 @@
+  Attribute getUnderlyingSpace() const { return memorySpaceAttr; }
+
+  /// Returns true if the underlying memory space is null.
+  bool isDefaultModel() const { return memorySpace == nullptr; }
```
The comment implies that this should be a check with `memorySpaceAttr`

Dinistro commented on mlir/include/mlir/Dialect/Ptr/IR/MemorySpace.h:
```diff
@@ -0,0 +1,161 @@
+
+protected:
+  /// Underlying memory space.
+  Attribute memorySpaceAttr{};
```
Nit: The naming is confusing. Maybe consider to use something like `underlyingMemorySpace`?

Dinistro reviewed (commented):
Very nice work. Dropped a few comments, but the design should be reviewed by someone else, I think.

gysit commented on mlir/lib/Dialect/Ptr/IR/PtrDialect.cpp:
```diff
@@ -39,8 +39,77 @@ void PtrDialect::initialize() {
+  if (!isa<PtrType>(ptrInfo.first))
+    /// The pointer-like operand is invalid.
+    return op ? op->emitError("invalid ptr-like type") : failure();
+  // Check shape validity.
```
The shape comparison seems to be duplicated here and above? I would try to factor this into a function? Maybe `getVecOrScalarInfo` could then just return the element type which seems to be what the remaining checks use. That way there is no need to factor out constants for -1 / -2.

joker-eph commented:
> Note: Ignore the first commit, that's being reviewed in https://github.com/llvm/llvm-project/pull/86860 .

How do we do that if you append more commits?

fabianmcg commented:
> How do we do that if you append more commits?

My plan was:
If the base is modified with new commits, ignore those changes if don't affect this PR, otherwise update the branch locally and force push. (It's unnecessary overhead, but I haven't migrated to `spr` and I think I saw a post on Discourse discouraging users from manually doing stack PRs)

If I add new commits to this PR, it shouldn't be an issue as the new commits will be on top.

Once, the dependent PR is approved and merged, I'll remove the first commit locally and then force push.

joker-eph commented:
> I think I saw a post on Discourse discouraging users from manually doing stack PRs)

I'd be curious where? I'm only doing manual stack PRs.

joker-eph commented on mlir/include/mlir/Dialect/Ptr/IR/MemorySpace.h:
```diff
@@ -0,0 +1,161 @@
+    if (auto intAttr = llvm::dyn_cast_or_null<IntegerAttr>(memorySpaceAttr))
+      return intAttr.getInt();
+    return 0;
+  }
```
Why is this needed/desirable?

fabianmcg replied:
I added it so it exposed a similar method to the one found in `memref` (`getMemorySpaceAsInt`), however, I missed the deprecated note for the method `memref`. I'll remove it.

joker-eph commented on mlir/include/mlir/Dialect/Ptr/IR/MemorySpace.h:
```diff
@@ -0,0 +1,161 @@
+namespace mlir {
+namespace ptr {
+/// This class wraps the `MemorySpaceAttrInterface` interface, providing a safe
+/// mechanism to specify the default behavior assumed by the ptr dialect.
```
Interfaces support default implementation in their specification, so you probably should flesh out the description here.

fabianmcg replied:
I'll change the description.

joker-eph commented on mlir/include/mlir/Dialect/Ptr/IR/MemorySpace.h:
```diff
@@ -0,0 +1,161 @@
+/// Compatible types are:
+/// Vectors of rank 1, or scalars of `ptr` type.
+LogicalResult isValidAddrSpaceCastImpl(Type tgt, Type src,
+                                       Operation *diagnosticOp);
```
In general we don't pass an operation for diagnostic, instead a `function_ref<InFlightDiagnostic()> emitError` works well.

joker-eph commented on mlir/include/mlir/Dialect/Ptr/IR/MemorySpace.h:
```diff
@@ -0,0 +1,161 @@
+/// IntLikeTy: Vectors of rank 1, or scalars of integer types or `index` type.
+/// PtrLikeTy: Vectors of rank 1, or scalars of `ptr` type.
+LogicalResult isValidPtrIntCastImpl(Type intLikeTy, Type ptrLikeTy,
+                                    Operation *diagnosticOp);
```
This seems to hardcode a bit of things around specific types, does it really belong here? What does it mean for the extensibility of the dialect?

fabianmcg replied:
That's just the default implementation used by `isValidAddrSpaceCast` . If an users wants to allow a different set of types, then they need to override the method.

fabianmcg commented:
The last commit addressed reviewer comments, resolve merge conflicts, and more importantly, it made `MemorySpaceAttrInterface` mandatory in `!ptr.ptr`. This decision to make it mandatory was made to improve semantics. In a future patch I'll add a default memory space for the ptr dialect.

fabianmcg commented:
Ping for review.

Dinistro commented:
I'll try to allocate time for this tomorrow, currently afk.
Side note: I suggest to always re-request reviews. This ensures that PRs show up as having an open review request.

Dinistro commented on mlir/include/mlir/Dialect/Ptr/IR/MemorySpaceInterfaces.td:
```diff
@@ -0,0 +1,162 @@
+      /*returnType=*/  "::mlir::ptr::MemorySpaceAttrInterface",
+      /*methodName=*/  "getDefaultMemorySpace",
+      /*args=*/        (ins),
+      /*methodBody=*/  [{}],
+      /*defaultImpl=*/ [{}]
```
IIRC, these do not need to be provided if they are empty.

Dinistro commented on mlir/include/mlir/Dialect/Ptr/IR/MemorySpaceInterfaces.td:
```diff
@@ -0,0 +1,162 @@
+    InterfaceMethod<
+      /*desc=*/        [{
+        This method checks if it's valid to perform a `ptrtoint` or `inttoptr`
+        op in the memory space. `CastValidity::InvalidSourceType` always refers
+        to the 'ptr-like' type and `CastValidity::InvalidTargetType` always
```
Is `CastValidity` a remnant from an older version? I can only see it here.

fabianmcg replied:
Thanks for catching this, it was a remnant.

Dinistro reviewed (approved):
Very nice work, and sorry for the delayed review. LGTM, but please give others a day or two to also take a look.

gysit commented on mlir/include/mlir/Dialect/Ptr/IR/MemorySpaceInterfaces.td:
```diff
@@ -0,0 +1,162 @@
+    InterfaceMethod<
+      /*desc=*/        "Returns the memory space of the op.",
+      /*returnType=*/  "::mlir::ptr::MemorySpaceAttrInterface",
+      /*methodName=*/  "getMemorySpace",
```
Would it make sense this function takes an (optional) pointer as attribute for operations that have multiple pointer operands that may work on different memory spaces? E.g. a copy operation that moves data between host and device? Although maybe we can postpone this to later once there is a use case...

fabianmcg replied:
We could add an arg for querying which address space (source or target) to return. However, I always thought of the target argument as the one establishing the semantics of the op.

gysit commented on mlir/include/mlir/Dialect/Ptr/IR/PtrEnums.td:
```diff
@@ -0,0 +1,69 @@
+include "mlir/IR/EnumAttr.td"
+
+//===----------------------------------------------------------------------===//
+// Atomic binary op enum attribute
```
```suggestion
// Atomic binary op enum attribute.
```
ultra nit:

gysit commented on mlir/include/mlir/Dialect/Ptr/IR/PtrEnums.td:
```diff
@@ -0,0 +1,69 @@
+}
+
+//===----------------------------------------------------------------------===//
+// Atomic ordering enum attribute
```
```suggestion
// Atomic ordering enum attribute.
```
ultra nit:

gysit commented on mlir/test/lib/Dialect/Test/TestAttributes.cpp:
```diff
@@ -316,6 +316,54 @@ static ParseResult parseCustomFloatAttr(AsmParser &p, StringAttr &typeStrAttr,
+LogicalResult TestConstMemorySpaceAttr::isValidStore(
+    Type type, mlir::ptr::AtomicOrdering ordering, IntegerAttr alignment,
+    function_ref<InFlightDiagnostic()> emitError) const {
+  return emitError ? (emitError() << "memory space is read-only") : failure();
```
Would it be possible to test these errors or is this only possible once you have a actual operations using it?

fabianmcg replied:
Properly testing, yes we need at least one op (next patch).
Minimal testing, I could add an unit test.

gysit reviewed (approved):
LGTM from my side.

fabianmcg committed:
address reviewer comments

joker-eph commented on mlir/include/mlir/Dialect/Ptr/IR/MemorySpaceInterfaces.td:
```diff
@@ -0,0 +1,142 @@
+      /*returnType=*/  "::mlir::ptr::MemorySpaceAttrInterface",
+      /*methodName=*/  "getMemorySpace"
+      >,
+  ];
```
It's not clear to me what it means for an op to have a memory space?

fabianmcg replied:
This is an internal interface to the ptr dialect to quickly identify the primary memory space being operated on. What about changing the method name to `getOperandMemorySpace` or `getPrimaryMemorySpace`?

joker-eph replied:
How will this be useful generically? What about a memcpy between memory spaces?

fabianmcg replied:
I can remove the interface.
This interface was mostly for convenience, so instead of going `getAddr().getType().getMemorySpace` one would call `getMemorySpace`, but there's no need to hide that behind an interface.

joker-eph commented on mlir/include/mlir/Dialect/Ptr/IR/MemorySpaceInterfaces.td:
```diff
@@ -0,0 +1,142 @@
+      }],
+      /*returnType=*/  "::mlir::ptr::MemorySpaceAttrInterface",
+      /*methodName=*/  "getDefaultMemorySpace"
+    >,
```
Is this always relevant for a given memory space attribute to have an answer about what a "default memory space" is?

fabianmcg replied:
We currently require a default space for the data layout. There are 2 options:
1. Create a convention and say that a given fixed attribute is the default space, let's say `#ptr.default_space`
2. Allow collections of memory spaces to specify what is the default space.

1 has the drawback that `#llvm.address_space<0>` wouldn't be the default, so it has the potential to create issues with existing code.

That's why I chose option 2.

joker-eph replied:
> We currently require a default space for the data layout

Right, but I don't quite see why it is relevant to the attribute interface of every memory space attribute.

fabianmcg replied:
The reasoning behind it is: Let's suppose we have memory spaces `read_only` and `write_only`. The `write_only` space represents a fat pointer with a different DL spec.:
```mlir
dlti.dl_spec = #dlti.dl_spec<
  #dlti.dl_entry<!ptr.ptr<#test.read_only>, #ptr.spec<size = 32, abi = 32, preferred = 64>>,
  #dlti.dl_entry<!ptr.ptr<#test.write_only>, #ptr.spec<size = 128, abi = 128, preferred = 128>>
>
```
Currently a query to the DL with `!ptr.ptr<#test.write_only<1>>` returns:
```mlir
 #dlti.dl_entry<!ptr.ptr<#test.write_only>, #ptr.spec<size = 128, abi = 128, preferred = 128>>
```
If there's no default spec per attribute class, then queries to either `!ptr.ptr<#test.read_only<1>>` or `!ptr.ptr<#test.write_only<1>>` will be invalid. Hence, one would need to list all possible specs for a memory space in the DL, which is possible but not desirable.

So it's not intrinsically relevant for all attributes, but simplifies querying the DL. However, we can remove it.

An alternative is adding an additional interface: `MemorySpaceWithDefaultSpace`

joker-eph replied:
> Currently a query to the DL with !ptr.ptr<#test.write_only<1>> returns:

A query for what? I'm not following the flow still at the moment.

fabianmcg commented on mlir/include/mlir/Dialect/Ptr/IR/MemorySpaceInterfaces.td:
```diff
@@ -0,0 +1,142 @@
+  let description = [{
+    This interface defines a common API for interacting with the memory model of
+    a memory space and the operations in the pointer dialect, giving proper
+    semantical meaning to the ops.
```
I'll change the phrase. What about:
`giving precise semantics to the ops. For example, establishing whether it is valid to read from an address or not.`

joker-eph replied:
I guess I still see a disconnect between an attribute interface and op semantics.

fabianmcg replied:
For example, for a read only space, `ptr.store` will error at verification, thus, in the end the attribute influences the op. However, I'll come up with a different description.

fabianmcg committed:
address reviewer comments

joker-eph commented on mlir/include/mlir/Dialect/Ptr/IR/MemorySpaceInterfaces.td:
```diff
@@ -0,0 +1,126 @@
+      /*desc=*/        [{
+        This method checks if it's valid to perform an `addrspacecast` op
+        in the memory space.
+        Both types are expected to be vectors of rank 1, or scalars of `ptr`
```
What the vector of rank-1 here?

fabianmcg committed:
address reviewer comments

joker-eph commented on mlir/include/mlir/Dialect/Ptr/IR/MemorySpaceInterfaces.td:
```diff
@@ -0,0 +1,117 @@
+        with a specific type, alignment, and atomic ordering.
+        If `emitError` is non-null then the method is allowed to emit errors.
+      }],
+      /*returnType=*/  "::mlir::LogicalResult",
```
@fabianmcg : nitpicking here but methods that starts with `is` are predicates and should return `bool`: `LogicalResult` is a about method that can fail instead. Can you update?

fabianmcg replied:
I can update. But these methods can emit an error, remember they get a `::llvm::function_ref<::mlir::InFlightDiagnostic()> emitError` parameter.

Admittedly,  I would prefer these methods to return something like a `SuccesOrDiagnostic`, so users can have more control on how to handle the result. Or to decouple the diagnostic from the validity method.

joker-eph replied:
`FailureOr<bool>`?

fabianmcg replied:
That's a good immediate option, I'll change to that. But I think we are missing something similar to `llvm::Expected`, like:

```c++
template <typename T>
class DiagnosticOr : public std::optional<T> {
   operator LogicalResult () const;
   Diagnostic takeDiagnostic();
   void emitDiagnostic();
   T& getValue();
private:
  Diagnostic message;
};
```
Which would allow users to decide how to handle the diagnostic.

joker-eph replied:
I think we're trying to be more efficient by avoiding formatting a diagnostic when the user does not want it.

fabianmcg replied:
On success we shouldn't format a diagnostic, but I'm still not convinced about the approach.
I'll change to `FailureOr<bool>`, I'll revisit if I have a better idea.

joker-eph replied:
> On success we shouldn't format a diagnostic

I'm talking about the failing case where the user of the API would just discard the diagnostic.

fabianmcg replied:
I just remembered  the `emitError` is always optional and the user of the API can opt to pass a default constructed object, in which case the method cannot produce errors. Therefore, it already handles the case of discarding the diagnostic in a efficient manner.

I'll switch to `FailureOr<bool>`.
